definitions/types/azure-storage-account-endpoint.json

```
@@ -0,0 +1,48 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema",
+  "title": "Azure Storage Account Authentication",
+  "description": "Azure Storage Account endpoint authentication",
+  "additionalProperties": false,
+  "examples": [
+    "https://storageaccount.blob.core.windows.net/",
+    "http://storageaccount.file.core.windows.net/",
+    "abfs://filesystem.accountname.dfs.core.windows.net/"
+  ],
+  "oneOf": [
+    {
+      "type": "string",
+      "pattern": "((http|https):\/\/)[a-z0-9]+\\.blob\\.core\\.windows\\.net\/$",
+      "message": {
+        "pattern": "Must be a valid Azure Storage Account blob endpoint"
+      }
+    },
+    {
+      "type": "string",
+      "pattern": "^((http|https):\/\/)[a-z0-9]+\\.queue\\.core\\.windows\\.net\/$",
+      "message": {
+        "pattern": "Must be a valid Azure Storage Account queue endpoint"
+      }
+    },
+    {
+      "type": "string",
+      "pattern": "^((http|https):\/\/)[a-z0-9]+\\.table\\.core\\.windows\\.net\/$",
+      "message": {
+        "pattern": "Must be a valid Azure Storage Account table endpoint"
+      }
+    },
+    {
+      "type": "string",
+      "pattern": "^((http|https):\/\/)[a-z0-9]+\\.file\\.core\\.windows\\.net\/$",
+      "message": {
+        "pattern": "Must be a valid Azure Storage Account file share endpoint"
+      }
+    },
+    {
+      "type": "string",
+      "pattern": "^((abfs|abfss|http|https):\/\/)([a-zA-Z0-9]+(\\.[a-zA-Z0-9]+)?)\\.dfs\\.core\\.windows\\.net\/$",
+      "message": {
+        "pattern": "Must be a valid Azure Storage Account DFS endpoint"
+      }
+    }
```

We can combine the http/https ones, what do you think?

```suggestion
     {
       "type": "string",
       "pattern": "((http|https):\/\/)[a-z0-9]+\\.[blob|dfs|file|queue|table]+\\.core\\.windows\\.net\/$",
       "message": {
         "pattern": "Must be a valid Azure Storage Account endpoint"
       }
     },
     {
       "type": "string",
       "pattern": "^((abfs|abfss):\/\/)([a-zA-Z0-9]+(\\.[a-zA-Z0-9]+)?)\\.dfs\\.core\\.windows\\.net\/$",
       "message": {
         "pattern": "Must be a valid Azure Storage Account DFS endpoint"
       }
     }
```

---

Seems cleaner and fancy, but that would break Azure Data Lake Storage API endpoints because it's optional to add a DNS reference in the URL, also we can specify the error message based on the endpoint used.

---

I actually don't know how it JSONSchema will determine which `oneOf` failed in order to report the `pattern` message. Technically all of them failed.

Honestly, I'd be fine w/ shrinking this down to one pattern that matches all of them:

```
"^((abfs|abfss|http|https):\/\/)[a-zA-Z0-9]+(\\.[a-zA-Z0-9]+)+\\.windows\\.net\\/$"
```

Artifact validations are much less important than parameter validations. If we have `type` definitions that the user will see, we want those to be solid. Types like this that are only used in artifacts don't need as much validation since the user won't see them (unless/until we allow external references). And at that point its mostly just making sure what they enter passes basic validation.

---

I'm fixing the regex right now to be able to support all storage account endpoint types.